File: serial_lib/src/main/java/com/kongqw/serialportlibrary/stick/BaseStickPackageHelper.java
package com.kongqw.serialportlibrary.stick;

import android.os.SystemClock;

import com.kongqw.serialportlibrary.utils.SerialPortLogUtil;

import java.io.IOException;
import java.io.InputStream;

/**
 * The simplest thing to do is not to deal with sticky packets,
 * read directly and return as much as InputStream.available() reads
 */
public class BaseStickPackageHelper implements AbsStickPackageHelper {
    public BaseStickPackageHelper() {
    }

    @Override
    public byte[] execute(InputStream is) {
        try {
            int available = is.available();
            if (available > 0) {
                byte[] buffer = new byte[available];
                int size = is.read(buffer);
                if (size > 0) {
                    return buffer;
                }
                SerialPortLogUtil.d("BaseStickPackageHelper", "原始数据长度: " + buffer.length);
            } else {
                SystemClock.sleep(50); // 默认50ms间隔
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: serial_lib/src/main/java/com/cl/serialportlibrary/stick/BaseStickPackageHelper.kt
package com.cl.serialportlibrary.stick

import com.cl.serialportlibrary.utils.SerialPortLogUtil
import java.io.IOException
import java.io.InputStream

/**
 * 不处理黏包，直接返回输入流当前可读取的数据。
 */
open class BaseStickPackageHelper @JvmOverloads constructor(
    private val readIntervalMillis: Long = DEFAULT_READ_INTERVAL_MS,
) : AbsStickPackageHelper {

    init {
        require(readIntervalMillis > 0) { "readIntervalMillis must be greater than 0" }
    }

    override fun execute(inputStream: InputStream): ByteArray? {
        return try {
            val available = inputStream.available()
            if (available > 0) {
                val buffer = ByteArray(available)
                val size = inputStream.read(buffer)
                if (size > 0) {
                    if (size == buffer.size) buffer else buffer.copyOf(size)
                } else {
                    null
                }
            } else {
                Thread.sleep(readIntervalMillis)
                null
            }
        } catch (error: IOException) {
            SerialPortLogUtil.e(TAG, "读取原始串口数据失败", error)
            throw error
        }
    }

    private companion object {
        private const val TAG = "BaseStickPackageHelper"
        private const val DEFAULT_READ_INTERVAL_MS = 50L
    }
}
